### PR TITLE
feat(validation): posterior-vs-reference comparison metrics + quantile op (#301 PR 1)

### PR DESCRIPTION
**Pull request by jhuggins**
## Summary

PR 1 of #301 — the in-repo, dependency-light scoring core for **method
validation**: score a posterior approximation against a trusted reference
(mean/covariance accuracy and distributional quality). It is method-agnostic
infrastructure — used to validate and regression-test any inference method, and
as the scoring core beneath `probpipe-benchmark` (#302) — with the
calibrated-SGLD initiative (#303/#304) as one motivating consumer. It lives in a
separate `probpipe.validation` subpackage, distinct from the per-fit diagnostics
in #193 (which it will reuse for ESS/R-hat in later PRs).

## What's in this PR

**`probpipe.validation._comparison`** — plain, jit-compatible JAX scoring
functions (typed via a `DrawsLike` = array-or-`flat_samples` alias):

- **`Reference`** — analytic / long-NUTS / sandwich references behind one
  container (high-precision `(mean, cov)`, `draws`, and/or `score_fn`);
  `from_moments` validates the `(d,)` / `(d, d)` shapes.
- **`standardized_mean_error`** = `‖Σ_ref^{-1/2}(μ̂ − μ_ref)‖₂` — Mahalanobis
  mean error, in posterior-σ units.
- **`relative_cov_error`** = `‖I − Σ_ref^{-1/2} Σ̂ Σ_ref^{-1/2}‖₂` — operator norm
  of the symmetrically whitened covariance deviation, which equals the
  worst-direction variance-ratio error `max_d |1 − λ_d|`.
- **`std_ratios`**, **`sliced_wasserstein`** (random projections → *exact*
  order-statistic 1-D `W₂`, handling unequal sample sizes), **`mmd`** (unbiased
  RBF), **`ksd`** (IMQ kernel Stein discrepancy), and **`score_posterior`** (the
  aggregator / scorecard reused by the future `probpipe-benchmark`; skips
  metrics whose reference pieces are absent). The moment metrics check
  approximation/reference dimension compatibility.

**`quantile` op** — a new `SupportsQuantile` protocol + `quantile(dist, q)` op
(parallel to `mean`/`variance`/`cov`) + `RecordEmpiricalDistribution._quantile`,
weight-aware via the midpoint-CDF (Hazen, type-5) quantile used for both uniform
and non-uniform weights, so the estimator is continuous in the weights.
Registered in the workflow distribution-hint allowlist so the op keeps its
distribution argument; rejects `q ∉ [0, 1]` (eager, skipped under `jit`).

**Docs + changelog** — a new `docs/api/validation.md` page (wired into the nav),
the `quantile` op added to `docs/api/operations.md`, and a CHANGELOG entry.

## Tests

- `tests/validation/test_comparison.py` — each metric vs an independent baseline
  (closed form / known identity / clearly-separated case), with measured
  tolerances per STYLE_GUIDE §8.6; the optimal-transport metrics are
  cross-checked against **POT** (`ot.wasserstein_1d` for the exact 1-D `W₂`
  including unequal `n`, and `ot.sliced_wasserstein_distance` for the projection
  pipeline). Also covers the non-diagonal `relative_cov_error` case, the
  float-bandwidth `mmd` / `ksd` paths, jit-compatibility, input rejection
  (dimension mismatch / `< 2` draws / 3-D), distribution inputs, and the
  `score_posterior` skip / selection paths.
- `tests/core/test_quantile_op.py` — uniform (vs NumPy Hazen) and weighted (vs an
  analytic `√q` and an independent NumPy baseline) quantiles, scalar/vector `q`,
  multi-field, the single-field return type, and `TypeError` / out-of-range-`q`
  errors.
- Full `tests/core` + `tests/validation` pass (1391); `ruff` clean.

## Scope notes

- **No new runtime dependencies** — the core install stays light. **POT** (Python
  Optimal Transport) is added as a **dev-only** test baseline for the
  optimal-transport metrics. SBC + interval coverage are PR 2; the shared
  `conftest` + references + the NUTS / vanilla-SGLD validation suite are PR 3
  (per #301).
- **Distinct from #193** (`probpipe.diagnostics`): validation scores a *method*
  against a reference; diagnostics diagnose a single *fit*. This PR reuses none
  of #193 yet (only `flat_samples` + the ops already on `main`).

Part of #301.

### CODE REVIEW

File: probpipe/validation/_comparison.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    _require(ref, "mean", "cov")
+    mu_hat = _as_draws(approx).mean(axis=0)
+    diff = mu_hat - ref.mean
+    chol = jnp.linalg.cholesky(ref.cov)
+    z = jax.scipy.linalg.solve_triangular(chol, diff, lower=True)
+    return jnp.linalg.norm(z)
+
+
+def relative_cov_error(approx: Any, ref: Reference) -> Array:
+    r"""Operator-norm covariance error: ``‖I − Σ_ref⁻¹ Σ̂‖₂``.
```
Comment (SoymilkWinsAgain) [inline probpipe/validation/_comparison.py]: The implementation of `relative_cov_error` is inconsistent with the description of "whitened covariance deviation / worst-direction variance-ratio error".

In `_comparison.py`, the current calculation is $\Sigma_{ref}^{-1} \Sigma_{hat}$, followed by $I - ...$. This matrix is ​​typically not symmetric: if the index semantics are the whitened covariance error in reference covariance coordinates, $L^{-1} \Sigma_{hat} L^{-T}$ should be calculated before taking $|I - ...|$. Current tests only cover unit/diagonal commutative scenarios and fail to capture this difference. It is recommended to modify it to a symmetric whitening form and add a test for a non-diagonal $\Sigma_{ref}$.

File: probpipe/core/ops.py [hunk 2]
```diff
@@ -252,6 +254,23 @@ def cov(dist: SupportsCovariance) -> Array:
...
+@workflow_function
+def quantile(dist: SupportsQuantile, q: Any) -> Any:
+    """Compute quantile(s) of ``X ~ dist`` at probability level(s) ``q``.
+
+    ``q`` is a scalar or array of probabilities in ``[0, 1]``; the return is
+    ``T``-shaped per field (finite-sample distributions return the weight-aware
+    empirical quantile via ``_quantile``).
+
+    Requires the distribution to implement :class:`SupportsQuantile`.
+    """
```
Comment (SoymilkWinsAgain) [inline probpipe/core/ops.py:269]: Should add check if q is in [0, 1].

**Review by SoymilkWinsAgain — CHANGES_REQUESTED** [review]
Also from CodeX:

The statement in the PR description that "uniform weights reduce to jnp.quantile" is outdated/inaccurate.

The actual implementation uses Hazen/type-5 midpoint-CDF, and testing is performed using `np.quantile(..., method="hazen")`. It is recommended to update the PR description; ideally, the finite-sample quantile convention should also be explicitly stated in the user-visible docstring for quantile.

File: probpipe/validation/_comparison.py (hunk 1 above)
Comment (jhuggins) [reply probpipe/validation/_comparison.py]: Thanks — good catch on the diagonal-only coverage. Switched to the symmetric whitened form `‖I − Σ_ref^{-1/2} Σ̂ Σ_ref^{-1/2}‖₂` (computed as `L⁻¹ Σ̂ L⁻ᵀ` via two triangular solves, `L = chol(Σ_ref)`) in f7c50a2. With the whitened matrix symmetric, the operator norm now **equals** the worst-direction variance-ratio deviation `max_d |1 − λ_d|` exactly, rather than the non-symmetry-inflated upper bound the previous form gave, and the docstring is updated to match. Added `test_relative_cov_error_non_diagonal`, which checks this against an independent NumPy eigendecomposition on a non-commuting `Σ̂`.

Two clarifications for the record:

- The previous `‖I − Σ_ref⁻¹ Σ̂‖₂` wasn't actually incorrect — it's a valid discrepancy and a genuine upper bound on the variance-ratio error, with equality exactly when `Σ̂` and `Σ_ref` commute (which is why the diagonal tests agreed). The symmetric form is just the cleaner, exact quantity, so we've adopted it.
- The left/right multiplication order doesn't change the value: `Σ̂ Σ_ref⁻¹ = (Σ_ref⁻¹ Σ̂)ᵀ`, so the operator norm is identical either way.

File: probpipe/core/ops.py (hunk 2 above)
Comment (jhuggins) [reply probpipe/core/ops.py:269]: Done in 5e9fcb7 — `quantile` now raises `ValueError` when a concrete `q` falls outside `[0, 1]`. The check is guarded so it is skipped when `q` is traced (e.g. under `jit`), keeping the op jit-compatible. Covered by `test_raises_on_out_of_range_q`.

File: probpipe/validation/_comparison.py (hunk 1 above)
Comment (SoymilkWinsAgain) [reply probpipe/validation/_comparison.py]: I'm not sure if this will cause some minor deviations in numerical calculations. This kind of problem can sometimes occur when using low-precision matrix calculations (though not necessarily here), and I've encountered similar issues when using low precision in `PyTorch`. Perhaps the docstrings should at least be consistent?

**Review by SoymilkWinsAgain — APPROVED** [review]
Everything looks fine!
